README.adoc

```diff
@@ -1,5 +1,174 @@
+    actorTypeBased: true
+----
+
+* `templates`: This defines a list of templates.
+
+* `name`: This is like an id for the template. Based on this, we can override the template later.
+
+* `pathExpression`: This specifies where the generated file is located and under what name it will be generated. We use spring expression language (springEL), because handlebars only support string output.
+
+* `templateName`: The location of the handlebars file relative to the root.
+
+* `actorTypeBased`: When actorTypeBased template used, the template called for all actor types and the actorType variable is defined.
+
+[source,java]
+----
+//if the actorTypeBased is true, we need to defined in the EsmGeneratorParameter
+
+    @Builder.Default
+    Predicate<ActorType> actorTypePredicate = a -> true;
+
+    @NonNull
+    Function<ActorType, File> actorTypeTargetDirectoryResolver;
+
+----
+Not in the example :
+
+* `factoryExpression`: Factory expression is used to create files. It returns a list of objects which are used as root context for the given handlebar template.
+
+* `templateContext`: Template context is used to put expression result to template variable. (Not used often because of the helpers)
+
+* `copy`: It can be used to copy a binary file. In this case the template file used as binary, no templating is performed.
+
+* `permission`: It can be used to add execution permission to files. For example bash files.
+
+* `exclude`: This parameter can be used in an override to exclude the given template from a generation. With this parameter only the `name` is effective.
+
+=== GeneratorModel
+
+The `GeneratorModel` Java class loads the Handlebars files from the base directory and overrides them with the files from the override directory if it is necessary.
+The class contains a global permission parameter that adds a global permission to all of the Handlebars files.
+(The `GeneratorTemplate` Java class, contained within the GeneratorModel class, includes all of the YAML representations listed earlier. The GeneratorModel maps the YAML properties to the GeneratorTemplate.)
+
+=== Generator Maven Project
+
+https://github.com/BlackBeltTechnology/judo-meta-esm/tree/develop/generator-maven-plugin#readme[This] link contains the description of the generator maven plugin.
+
+=== Helper registration
+
+We don't register helpers manually. To register a java helper class use the `@TemplateHelper` annotation.
+
+IMPORTANT: The Java class method must be static.
+
+If you inherit your helper class from the `StaticMethodValueResolver` class, you can refer to its methods using dot notation.
+
+The helpers are available in springEL too with # prefix.
+
+=== Self meaning in different situations
+
+The self mean different thinks in the springEL and Handlebars depend on the YAML properties:
+
+- If actorTypeBased is enabled, then #self refers to the actor.
+
+- If we have a `factoryExpression`, then #self refers to the result of the factory expression.
```

Comment:
The result is a collection. The self is the current iterated element of collection